File: src/lib/blockchain.ts
/**
 * SecureLand Blockchain Engine
 * SHA-256 Hash Chain with Proof-of-Work
 * Runs client-side, stores blocks in Firestore
 */

import { db } from "./firebase";
import {
    doc,
    setDoc,
    getDoc,
    getDocs,
    collection,
    query,
    where,
    orderBy,
    updateDoc,
} from "firebase/firestore";

// =============================================
// SHA-256 Helper (Web Crypto API)
// =============================================
async function sha256(message: string): Promise<string> {
    const msgBuffer = new TextEncoder().encode(message);
    const hashBuffer = await crypto.subtle.digest("SHA-256", msgBuffer);
    const hashArray = Array.from(new Uint8Array(hashBuffer));
    return hashArray.map((b) => b.toString(16).padStart(2, "0")).join("");
}

// =============================================
// Block Interfaces
// =============================================
export interface Block {
    index: number;
    timestamp: string;
    landId: string;
    ownerName: string;
    mobile: string;
    state: string;
    location: string;
    area: number;
    coordinates: { lat: number; lng: number }[];
    previousHash: string;
    nonce: number;
    hash: string;
    transactionType: string;
}

export interface ChainMetadata {
    totalBlocks: number;
    lastBlockIndex: number;
    lastBlockHash: string;
    chainCreated: string;
    lastUpdated: string;
}

// =============================================
// Hash Generation
// =============================================
async function generateBlockHash(block: Partial<Block>): Promise<string> {
    const data = JSON.stringify({
        index: block.index,
        timestamp: block.timestamp,
        landId: block.landId,
        ownerName: block.ownerName,
        mobile: block.mobile,
        state: block.state,
        location: block.location,
        area: block.area,
        coordinates: block.coordinates,
        previousHash: block.previousHash,
        nonce: block.nonce,
    });
    return sha256(data);
}

// =============================================
// Proof-of-Work Mining (2 leading zeros)
// =============================================
async function mineBlock(
    blockData: Partial<Block>
): Promise<{ nonce: number; hash: string }> {
    let nonce = 0;
    let hash = "";
    const difficulty = "00";

    do {
        nonce++;
        blockData.nonce = nonce;
        hash = await generateBlockHash(blockData);
    } while (!hash.startsWith(difficulty) && nonce < 100000);

    return { nonce, hash };
}

// =============================================
// Genesis Block
// =============================================
async function ensureGenesisBlock(): Promise<Block> {
    const genesisRef = doc(db, "blockchain", "block_0");
    const genesisDoc = await getDoc(genesisRef);

    if (!genesisDoc.exists()) {
        const genesisBlock: Block = {
            index: 0,
            timestamp: new Date().toISOString(),
            landId: "GENESIS",
            ownerName: "SecureLand System",
            mobile: "0000000000",
            state: "System",
            location: "SecureLand Blockchain Network",
            area: 0,
            coordinates: [],
            previousHash:
                "0000000000000000000000000000000000000000000000000000000000000000",
            nonce: 0,
            hash: "",
            transactionType: "GENESIS",
        };

        genesisBlock.hash = await generateBlockHash(genesisBlock);
        await setDoc(genesisRef, genesisBlock);

        // Set chain metadata
        await setDoc(doc(db, "blockchain", "_metadata"), {
            totalBlocks: 1,
            lastBlockIndex: 0,
            lastBlockHash: genesisBlock.hash,
            chainCreated: genesisBlock.timestamp,
            lastUpdated: genesisBlock.timestamp,
        });

        console.log("⛓️ Genesis block created:", genesisBlock.hash);
        return genesisBlock;
    }

    return genesisDoc.data() as Block;
}

// =============================================
// PUBLIC API: Register Land on Blockchain
// =============================================
export async function registerOnBlockchain(landData: {
    landId: string;
    ownerName: string;
    mobile: string;
    state: string;
    location: string;
    area: number;
    coordinates: { lat: number; lng: number }[];
}): Promise<Block> {
    // 1. Ensure genesis block exists
    await ensureGenesisBlock();

    // 2. Get chain metadata
    const metaRef = doc(db, "blockchain", "_metadata");
    const metaDoc = await getDoc(metaRef);
    const metadata = metaDoc.data() as ChainMetadata;

    const newIndex = metadata.lastBlockIndex + 1;

    // 3. Create block data
    const blockData: Partial<Block> = {
        index: newIndex,
        timestamp: new Date().toISOString(),
        landId: landData.landId,
        ownerName: landData.ownerName,
        mobile: landData.mobile,
        state: landData.state,
        location: landData.location,
        area: landData.area || 0,
        coordinates: landData.coordinates || [],
        previousHash: metadata.lastBlockHash,
        nonce: 0,
        transactionType: "LAND_REGISTRATION",
    };

    // 4. Mine (Proof-of-Work)
    const { nonce, hash } = await mineBlock(blockData);
    blockData.nonce = nonce;
    blockData.hash = hash;

    // 5. Store block in Firestore
    const blockRef = doc(db, "blockchain", `block_${newIndex}`);
    await setDoc(blockRef, blockData);

    // 6. Update metadata
    await updateDoc(metaRef, {
        totalBlocks: newIndex + 1,
        lastBlockIndex: newIndex,
        lastBlockHash: hash,
        lastUpdated: blockData.timestamp,
    });

    // 7. Update Digital Twin with blockchain info
    const twinRef = doc(db, "digitalTwins", landData.landId);
    await updateDoc(twinRef, {
        blockchainHash: hash,
        blockIndex: newIndex,
        previousBlockHash: metadata.lastBlockHash,
        blockNonce: nonce,
        blockchainTimestamp: blockData.timestamp,
        blockchainVerified: true,
    });

    console.log(`⛏️ Block #${newIndex} mined: ${hash} (nonce: ${nonce})`);
    return blockData as Block;
}

// =============================================
// PUBLIC API: Verify a Land Record
// =============================================
export async function verifyLandOnBlockchain(landId: string): Promise<{
    verified: boolean;
    block: Block | null;
    hashValid: boolean;
    chainValid: boolean;
}> {
    // Find block with this landId
    const q = query(
        collection(db, "blockchain"),
        where("landId", "==", landId)
    );
    const snapshot = await getDocs(q);

    if (snapshot.empty) {
        return { verified: false, block: null, hashValid: false, chainValid: false };
    }

    const block = snapshot.docs[0].data() as Block;

    // Verify hash
    const recalculatedHash = await generateBlockHash(block);
    const hashValid = recalculatedHash === block.hash;

    // Verify chain link
    let chainValid = true;
    if (block.index > 0) {
        const prevDoc = await getDoc(
            doc(db, "blockchain", `block_${block.index - 1}`)
        );
        if (prevDoc.exists()) {
            const prevBlock = prevDoc.data() as Block;
            chainValid = block.previousHash === prevBlock.hash;
        }
    }

    return {
        verified: hashValid && chainValid,
        block,
        hashValid,
        chainValid,
    };
}

// =============================================
// PUBLIC API: Get Full Chain
// =============================================
export async function getBlockchain(): Promise<Block[]> {
    const q = query(collection(db, "blockchain"), orderBy("index", "asc"));
    const snapshot = await getDocs(q);

    const blocks: Block[] = [];
    snapshot.forEach((docSnap) => {
        const data = docSnap.data();
        if (docSnap.id !== "_metadata" && data.index !== undefined) {
            blocks.push(data as Block);
        }
    });

    return blocks;
}

// =============================================
// PUBLIC API: Get Chain Stats
// =============================================
export async function getBlockchainStats(): Promise<ChainMetadata | null> {
    const metaDoc = await getDoc(doc(db, "blockchain", "_metadata"));
    if (!metaDoc.exists()) {
        await ensureGenesisBlock();
        const newMeta = await getDoc(doc(db, "blockchain", "_metadata"));
        return newMeta.data() as ChainMetadata;
    }
    return metaDoc.data() as ChainMetadata;
}

// =============================================
// PUBLIC API: Validate Entire Chain
// =============================================
export async function validateBlockchain(): Promise<{
    valid: boolean;
    totalBlocks: number;
    errors: string[];
}> {
    const metaDoc = await getDoc(doc(db, "blockchain", "_metadata"));
    if (!metaDoc.exists()) {
        return { valid: false, totalBlocks: 0, errors: ["No blockchain found"] };
    }

    const metadata = metaDoc.data() as ChainMetadata;
    const errors: string[] = [];

    for (let i = 0; i < metadata.totalBlocks; i++) {
        const blockDoc = await getDoc(doc(db, "blockchain", `block_${i}`));
        if (!blockDoc.exists()) {
            errors.push(`Block #${i} missing`);
            continue;
        }

        const block = blockDoc.data() as Block;

        // Verify hash
        const recalculatedHash = await generateBlockHash(block);
        if (recalculatedHash !== block.hash) {
            errors.push(`Block #${i}: Hash mismatch — DATA TAMPERED`);
        }

        // Verify chain (skip genesis)
        if (i > 0) {
            const prevDoc = await getDoc(doc(db, "blockchain", `block_${i - 1}`));
            if (prevDoc.exists()) {
                const prevBlock = prevDoc.data() as Block;
                if (block.previousHash !== prevBlock.hash) {
                    errors.push(`Block #${i}: Chain broken — previous hash mismatch`);
                }
            }
        }
    }

    return {
        valid: errors.length === 0,
        totalBlocks: metadata.totalBlocks,
        errors,
    };
}

File: backend/api/blockchain.py
"""
Blockchain Engine — SHA-256 Hash Chain with Proof-of-Work.
Runs server-side in Django, stores blocks in Firestore.
"""

import hashlib
import json
from datetime import datetime
from firebase_admin import firestore


def get_db():
    return firestore.client()


def generate_block_hash(block_data: dict) -> str:
    """Generate SHA-256 hash of a block."""
    data = json.dumps({
        "index": block_data.get("index"),
        "timestamp": block_data.get("timestamp"),
        "landId": block_data.get("landId"),
        "ownerName": block_data.get("ownerName"),
        "mobile": block_data.get("mobile"),
        "state": block_data.get("state"),
        "location": block_data.get("location"),
        "area": block_data.get("area"),
        "coordinates": block_data.get("coordinates"),
        "previousHash": block_data.get("previousHash"),
        "nonce": block_data.get("nonce"),
    }, sort_keys=False)
    return hashlib.sha256(data.encode()).hexdigest()


def mine_block(block_data: dict, difficulty: str = "00") -> tuple:
    """Proof-of-Work: find nonce that produces hash starting with difficulty prefix."""
    nonce = 0
    while nonce < 100000:
        nonce += 1
        block_data["nonce"] = nonce
        hash_val = generate_block_hash(block_data)
        if hash_val.startswith(difficulty):
            return nonce, hash_val
    return nonce, generate_block_hash(block_data)


def ensure_genesis_block():
    """Create genesis block if it doesn't exist."""
    db = get_db()
    genesis_ref = db.collection("blockchain").document("block_0")
    genesis_doc = genesis_ref.get()

    if not genesis_doc.exists:
        genesis_block = {
            "index": 0,
            "timestamp": datetime.utcnow().isoformat() + "Z",
            "landId": "GENESIS",
            "ownerName": "SecureLand System",
            "mobile": "0000000000",
            "state": "System",
            "location": "SecureLand Blockchain Network",
            "area": 0,
            "coordinates": [],
            "previousHash": "0" * 64,
            "nonce": 0,
            "hash": "",
            "transactionType": "GENESIS",
        }
        genesis_block["hash"] = generate_block_hash(genesis_block)
        genesis_ref.set(genesis_block)

        db.collection("blockchain").document("_metadata").set({
            "totalBlocks": 1,
            "lastBlockIndex": 0,
            "lastBlockHash": genesis_block["hash"],
            "chainCreated": genesis_block["timestamp"],
            "lastUpdated": genesis_block["timestamp"],
        })
        return genesis_block

    return genesis_doc.to_dict()


def register_on_blockchain(land_data: dict) -> dict:
    """Add a new block to the blockchain for a land registration."""
    db = get_db()
    ensure_genesis_block()

    meta_ref = db.collection("blockchain").document("_metadata")
    meta_doc = meta_ref.get()
    metadata = meta_doc.to_dict()

    new_index = metadata["lastBlockIndex"] + 1

    block_data = {
        "index": new_index,
        "timestamp": datetime.utcnow().isoformat() + "Z",
        "landId": land_data["landId"],
        "ownerName": land_data.get("ownerName", ""),
        "mobile": land_data.get("mobile", ""),
        "state": land_data.get("state", ""),
        "location": land_data.get("location", ""),
        "area": land_data.get("area", 0),
        "coordinates": land_data.get("coordinates", []),
        "previousHash": metadata["lastBlockHash"],
        "nonce": 0,
        "transactionType": "LAND_REGISTRATION",
    }

    nonce, hash_val = mine_block(block_data)
    block_data["nonce"] = nonce
    block_data["hash"] = hash_val

    # Store block
    db.collection("blockchain").document(f"block_{new_index}").set(block_data)

    # Update metadata
    meta_ref.update({
        "totalBlocks": new_index + 1,
        "lastBlockIndex": new_index,
        "lastBlockHash": hash_val,
        "lastUpdated": block_data["timestamp"],
    })

    # Update Digital Twin
    twin_ref = db.collection("digitalTwins").document(land_data["landId"])
    twin_ref.update({
        "blockchainHash": hash_val,
        "blockIndex": new_index,
        "previousBlockHash": metadata["lastBlockHash"],
        "blockNonce": nonce,
        "blockchainTimestamp": block_data["timestamp"],
        "blockchainVerified": True,
    })

    return block_data


def verify_land_on_blockchain(land_id: str) -> dict:
    """Verify a land record on the blockchain."""
    db = get_db()
    query = db.collection("blockchain").where("landId", "==", land_id).limit(1).get()

    if not query:
        return {"verified": False, "block": None, "hashValid": False, "chainValid": False}

    block = query[0].to_dict()
    recalculated = generate_block_hash(block)
    hash_valid = recalculated == block.get("hash")

    chain_valid = True
    if block["index"] > 0:
        prev_doc = db.collection("blockchain").document(f"block_{block['index'] - 1}").get()
        if prev_doc.exists:
            chain_valid = block["previousHash"] == prev_doc.to_dict()["hash"]

    return {
        "verified": hash_valid and chain_valid,
        "block": block,
        "hashValid": hash_valid,
        "chainValid": chain_valid,
    }


def get_full_chain() -> list:
    """Get the full blockchain."""
    db = get_db()
    docs = db.collection("blockchain").order_by("index").get()
    blocks = []
    for doc in docs:
        data = doc.to_dict()
        if doc.id != "_metadata" and "index" in data:
            blocks.append(data)
    return blocks


def validate_chain() -> dict:
    """Validate the entire blockchain integrity."""
    db = get_db()
    meta_doc = db.collection("blockchain").document("_metadata").get()
    if not meta_doc.exists:
        return {"valid": False, "totalBlocks": 0, "errors": ["No blockchain found"]}

    metadata = meta_doc.to_dict()
    errors = []

    for i in range(metadata["totalBlocks"]):
        block_doc = db.collection("blockchain").document(f"block_{i}").get()
        if not block_doc.exists:
            errors.append(f"Block #{i} missing")
            continue

        block = block_doc.to_dict()
        recalculated = generate_block_hash(block)
        if recalculated != block.get("hash"):
            errors.append(f"Block #{i}: Hash mismatch — DATA TAMPERED")

        if i > 0:
            prev_doc = db.collection("blockchain").document(f"block_{i - 1}").get()
            if prev_doc.exists:
                if block["previousHash"] != prev_doc.to_dict()["hash"]:
                    errors.append(f"Block #{i}: Chain broken")

    return {
        "valid": len(errors) == 0,
        "totalBlocks": metadata["totalBlocks"],
        "errors": errors,
    }
